### PR TITLE
Add spill readahead support

### PR DESCRIPTION
**Pull request by xiaoxmeng**
We add read-ahead support when filesystem used for spill support async read, like Meta internal
remote storage. This help pipeline the spill data read and spill data processing. With a medium sized
Meta internal query, the query execution time has been reduced from 12 mins
(20240607_225722_00001_tvvdk) to 7 mins (20240608_030802_00001_xjs2q)

### CODE REVIEW

File: velox/exec/SpillFile.cpp [hunk 1]
```diff
@@ -28,17 +28,85 @@ namespace {
...
+}
+
+SpillInputStream::~SpillInputStream() {
+  if (!readaWait_.valid()) {
+    return;
+  }
+  try {
+    readaWait_.wait();
+  } catch (const std::exception& ex) {
+    // ignore any prefetch error when query has failed.
```
Comment (duanmeng) [inline velox/exec/SpillFile.cpp:59]: s/ignore/Ignore

**Review by duanmeng — APPROVED** [review]
One nit, thanks.

File: velox/exec/SpillFile.h [hunk 2]
```diff
@@ -209,30 +209,65 @@ class SpillInputStream : public ByteInputStream {
...
  private:
   void updateSpillStats(uint64_t readBytes, uint64_t readTimeUs) const;
+
   void next(bool throwIfPastEnd) override;
 
+  // Issues readahead if underlying fs supports async mode read.
+  //
+  // TODO: we might consider to use AsyncSource to support read-ahead on
+  // filesystem which doesn't support async mode read.
+  void maybeIssueReadahead();
```
Comment (tanjialiang) [inline velox/exec/SpillFile.h:232]: s/maybeIssueReadahead/maybeIssueReadAhead

File: velox/exec/SpillFile.h [hunk 3]
```diff
@@ -209,30 +209,65 @@ class SpillInputStream : public ByteInputStream {
...
-    return offset_ >= size_ && ranges()[0].position >= ranges()[0].size;
+    return offset_ >= fileSize_ && ranges()[0].position >= ranges()[0].size;
   }
 
  private:
   void updateSpillStats(uint64_t readBytes, uint64_t readTimeUs) const;
+
   void next(bool throwIfPastEnd) override;
 
+  // Issues readahead if underlying fs supports async mode read.
```
Comment (tanjialiang) [inline velox/exec/SpillFile.h:228]: s/fs/file system/

Comment (tanjialiang) [reply velox/exec/SpillFile.h:228]: s/async mode read/async read mode/

File: velox/exec/SpillFile.h [hunk 4]
```diff
@@ -209,30 +209,65 @@ class SpillInputStream : public ByteInputStream {
...
+  //
+  // TODO: we might consider to use AsyncSource to support read-ahead on
+  // filesystem which doesn't support async mode read.
+  void maybeIssueReadahead();
+
+  inline uint32_t bufferIndex() const {
+    return bufferIndex_;
+  }
```
Comment (tanjialiang) [inline velox/exec/SpillFile.h:236]: This is not needed.

File: velox/exec/SpillFile.cpp [hunk 5]
```diff
@@ -28,17 +28,85 @@ namespace {
...
-  {
-    MicrosecondTimer timer{&readTimeUs};
-    file_->pread(offset_, readBytes, buffer_->asMutable<char>());
+  if (readaWait_.valid()) {
+    {
+      MicrosecondTimer timer{&readTimeUs};
+      readBytes = std::move(readaWait_)
+                      .via(&folly::QueuedImmediateExecutor::instance())
+                      .wait()
+                      .value();
```
Comment (tanjialiang) [inline velox/exec/SpillFile.cpp:73]: .via().get()

**Review by tanjialiang — COMMENTED** [review]
Are we able to extend this to more than 2 buffers?

**Pull request closed without merging**